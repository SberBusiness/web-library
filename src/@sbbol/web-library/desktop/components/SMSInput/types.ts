import * as React from 'react';
import {ESMSInputSize} from './enums';
import {TestProps} from '@sbbol/web-library/desktop/common/types/CoreTypes';

/** Свойство компонента SMSInput. */
export interface ISMSInputProps extends React.HTMLAttributes<HTMLDivElement>, TestProps {
    /** Признак блокировки компонента. */
    disabled?: boolean;
    /** Признак блокировки кнопки запроса кода. */
    disabledRefresh?: boolean;
    /** Признак блокировки кнопки отправки кода. */
    disabledSubmit?: boolean;
    /** Признак наличия ошибки. */
    hasError?: boolean;
    /** Максимальное количество символов в инпуте. */
    maxLength?: number;
    /** Сообщение для тултипа. */
    message: string;
    /** Сообщение для тултипа во время обратного отсчета. */
    messageTicking: string;
    /** Обработчик повторного запроса кода. */
    onRefreshCode: (event?: React.MouseEvent) => void;
    /** Обработчик окончания обратного отсчета. */
    onRefreshTimeout?: () => void;
    /** Обработчик отправки кода. */
    onSubmitCode: (code: string, event?: React.MouseEvent | React.KeyboardEvent) => void;
    /** Размер поля. */
    size?: ESMSInputSize; // TODO: Сделать свойство обязательным в ближайшем мажорном релизе.
    /** Задержка повторной отправки СМС кода. */
    smsCountdownTime: number;
    /** Запускать обратный отсчет при маунте компонента. */
    startTimerOnMount?: boolean;
}

/** Состояние компонента SMSInput. */
export interface ISMSInputState {
    /** Введенный код. */
    code: string;
    /** Количество секунд, оставшееся до возможности повторного запроса СМС-кода. */
    smsCountdownTimeLeft: number;
}
