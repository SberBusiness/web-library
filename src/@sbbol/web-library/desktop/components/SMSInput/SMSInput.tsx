import * as React from 'react';
import {ISMSInputProps, ISMSInputState} from '@sbbol/web-library/desktop/components/SMSInput/types';
import {ESMSInputSize} from '@sbbol/web-library/desktop/components/SMSInput/enums';
import {Tooltip} from '@sbbol/web-library/desktop/components/Tooltip/Tooltip';
import {ETooltipAlign, ETooltipSize} from '@sbbol/web-library/desktop/components/Tooltip/enums';
import {TooltipBody} from '@sbbol/web-library/desktop/components/Tooltip/TooltipBody';
import {TooltipTarget} from '@sbbol/web-library/desktop/components/Tooltip/TooltipTarget';
import {Input} from '@sbbol/web-library/desktop/components/Input/Input';
import {classnames} from '@sbbol/web-library/common/utils/classnames/classnames';
import {RefreshButton} from '@sbbol/web-library/desktop/components/SMSInput/RefreshButton';
import {ArrowRightWhiteIcon20} from '@sbbol/web-library/desktop/common/icons/ArrowRightWhiteIcon20';
import {TestIds} from '@sbbol/web-library/common/dataTestIds/dataTestIds';
import {EVENT_KEY_CODES} from '@sbbol/web-library/desktop/utils/keyboard';

/** Возвращает CSS класс в зависимости от размера. */
const getSizeClassName = (size?: ESMSInputSize) => {
    switch (size) {
        case ESMSInputSize.MD:
            return 'cssClass[MD]';
        case ESMSInputSize.SM:
            return 'cssClass[SM]';
    }
};

/** Компонент для ввода СМС. */
export class SMSInput extends React.PureComponent<ISMSInputProps, ISMSInputState> {
    public static displayName = 'SMSInput';
    public regExp = new RegExp(/^[0-9]*$/);

    public static defaultProps = {
        maxLength: 8,
        size: ESMSInputSize.MD,
    };

    public state = {
        code: '',
        smsCountdownTimeLeft: 0,
    };

    private smsIntervalId: number | null = null;

    public componentDidMount(): void {
        if (this.props.startTimerOnMount) {
            this.startCountdown();
        }
    }

    public componentDidUpdate(prevProps: Readonly<ISMSInputProps>): void {
        if (prevProps.disabledSubmit && !this.props.disabledSubmit && this.props.hasError) {
            this.setState({code: ''});
        }
    }

    public componentWillUnmount(): void {
        if (this.smsIntervalId) {
            clearInterval(this.smsIntervalId);
        }
    }

    public render(): JSX.Element {
        const {
            className,
            disabled,
            disabledRefresh,
            disabledSubmit,
            hasError,
            message,
            messageTicking,
            onRefreshCode,
            onRefreshTimeout,
            onSubmitCode,
            size,
            smsCountdownTime,
            maxLength,
            placeholder,
            ...htmlDivAttributes
        } = this.props;
        const dataTestId = this.props['data-test-id'];
        const {code, smsCountdownTimeLeft} = this.state;

        const isSubmitDisabled = this.isSubmitDisabled();

        let percent = 1; // Проценты выражаются в долях единицы.
        if (smsCountdownTimeLeft && smsCountdownTime) {
            if (smsCountdownTimeLeft >= smsCountdownTime) {
                percent = 0;
            } else {
                percent = (smsCountdownTime - smsCountdownTimeLeft) / smsCountdownTime;
            }
        }

        const isSmsCountdownTicking = smsCountdownTimeLeft > 0;

        const isRefreshDisabled = (disabled && !hasError) || disabledRefresh || isSmsCountdownTicking;

        const classSMSInput = classnames(className, 'cssClass[globalSmsInput]');
        const classBtnSubmit = classnames('cssClass[btnSubmit]', getSizeClassName(size), {'cssClass[disabled]': isSubmitDisabled});
        const classBtnRefresh = classnames('cssClass[btnRefresh]', getSizeClassName(size), {'cssClass[disabled]': isRefreshDisabled});
        const classInput = classnames('cssClass[input]', getSizeClassName(size), {
            'cssClass[error]': Boolean(hasError),
            'cssClass[filled]': code !== '',
        });

        // Приводит оставшееся время к формату mm:ss
        const formattedTimeLeft = new Date(smsCountdownTimeLeft * 1000).toISOString().substr(14, 5);

        const smsMessage = isSmsCountdownTicking ? `${messageTicking} ${formattedTimeLeft}` : message;

        const placeholderText = placeholder || (hasError ? 'Неверный код' : 'Введите код');

        return (
            <div className={classSMSInput} {...htmlDivAttributes}>
                <Tooltip size={ETooltipSize.SM} toggleType="hover" alignTip={ETooltipAlign.START}>
                    <TooltipBody>{smsMessage}</TooltipBody>
                    <TooltipTarget>
                        <div
                            className={classBtnRefresh}
                            data-test-id={dataTestId && `${dataTestId}${TestIds.SMSInput.button_refresh}`}
                            role="button"
                            tabIndex={0}
                            {...(!isRefreshDisabled && {onClick: this.handleRefresh})}
                        >
                            <RefreshButton percent={percent} />
                        </div>
                    </TooltipTarget>
                </Tooltip>
                <Input
                    className={classInput}
                    value={code}
                    placeholder={placeholderText}
                    maxLength={maxLength}
                    autoComplete="off"
                    onChange={this.handleChangeCode}
                    onKeyDown={this.handleKeyDown}
                    disabled={disabled}
                    data-test-id={dataTestId && `${dataTestId}${TestIds.SMSInput.input}`}
                />
                <div
                    className={classBtnSubmit}
                    data-test-id={dataTestId && `${dataTestId}${TestIds.SMSInput.button_submit}`}
                    {...(!isSubmitDisabled && {onClick: this.handleSubmitCode})}
                >
                    <ArrowRightWhiteIcon20 />
                </div>
            </div>
        );
    }

    private isSubmitDisabled = () => {
        const {disabled, disabledSubmit} = this.props;
        const {code} = this.state;

        return disabled || disabledSubmit || code === '';
    };

    /**
     * Обработчик ввода sms-кода.
     * @param {React.ChangeEvent<HTMLInputElement>} e Событие.
     */
    private handleChangeCode = (e: React.ChangeEvent<HTMLInputElement>) => {
        const value = e.target.value;
        if (this.regExp.test(value)) {
            this.setState({code: value});
        }
    };

    /**
     * Обработчик отправки sms-кода.
     */
    private handleSubmitCode = (event: React.MouseEvent | React.KeyboardEvent) => {
        if (!this.isSubmitDisabled()) {
            this.props.onSubmitCode(this.state.code, event);
        }
    };

    /**
     * Обработчик нажатия клавиши Enter (для отправки СМС-кода).
     */
    private handleKeyDown = (event: React.KeyboardEvent<HTMLInputElement>) => {
        if (event.keyCode === EVENT_KEY_CODES.ENTER) {
            this.handleSubmitCode(event);
        }
    };

    /**
     * Обработчик запроса нового sms-кода.
     */
    private handleRefresh = (event: React.MouseEvent) => {
        this.props.onRefreshCode(event);
        this.setState({code: ''});
        this.startCountdown();
    };

    /**
     * Запуск обратного отсчета.
     */
    private startCountdown = () => {
        this.smsIntervalId = window.setInterval(this.intervalChange, 1000);
        this.setState({smsCountdownTimeLeft: this.props.smsCountdownTime});
    };

    /**
     * Ежесекундный коллбэк.
     */
    private intervalChange = () => {
        this.setState(
            (prevState) => ({smsCountdownTimeLeft: prevState.smsCountdownTimeLeft - 1}),
            () => {
                const timerExpired = this.state.smsCountdownTimeLeft <= 0;
                if (timerExpired) {
                    if (this.props.onRefreshTimeout) {
                        this.props.onRefreshTimeout();
                    }
                    clearInterval(this.smsIntervalId!);
                }
            }
        );
    };
}
