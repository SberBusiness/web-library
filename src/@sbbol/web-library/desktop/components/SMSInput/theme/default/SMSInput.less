@import '../../../../common/theme/default';

.globalSmsInput {
    display: inline-block;
    position: relative;
    width: 100%;

    .input {
        padding-left: 33px;
        padding-right: 33px;
        background-color: @color-snowy;
        border-radius: @radius-16;
        box-sizing: border-box;
        text-align: center;

        &.SM {
            height: 28px;
        }

        &:focus:enabled,
        &.filled:enabled {
            border-color: @color-graphite-01;
        }

        &.error,
        &.error.filled,
        &.error:focus {
            border-color: @color-ruby-02;
        }
    }

    .btnRefresh {
        position: absolute;
        cursor: pointer;
        left: 6px;
        outline: none;

        &.MD {
            padding-top: 6px;
        }

        &.SM {
            padding-top: 4px;
        }

        &.disabled {
            cursor: default;
        }

        svg {
            outline: none;
        }

        .light {
            fill: @color-asphalt-02;
        }

        .dark {
            fill: @color-graphite-02;
        }
    }

    .btnSubmit {
        cursor: pointer;
        position: absolute;
        right: 6px;
        width: 20px;
        height: 20px;
        border-radius: 50%;
        background-color: @color-malachite-02;

        &.MD {
            top: 6px;
        }

        &.SM {
            top: 4px;
        }

        &.disabled {
            cursor: default;
            background-color: @color-malachite-04;
        }
    }
}
