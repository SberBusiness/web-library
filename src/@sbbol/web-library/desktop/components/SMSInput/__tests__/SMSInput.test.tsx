import React from 'react';
import renderer from 'react-test-renderer';
import {SMSInput} from '../SMSInput';
import {allure} from '@jest/unit/allure-report';

jest.mock('@sberbusiness/icons/ClosetooltipSrvxIcon16', () => ({
    ClosetooltipSrvxIcon16: 'svg',
}));

describe.skip('SMSInput', () => {
    beforeEach(() => {
        allure.feature('SMSInput');
    });

    it('renders correctly', () => {
        const tree = renderer
            .create(
                <SMSInput
                    data-test-id="SMSInput"
                    message="Запросить новый код"
                    messageTicking="Запросить новый код через"
                    onRefreshCode={jest.fn()}
                    onSubmitCode={jest.fn()}
                    smsCountdownTime={0}
                />
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});
