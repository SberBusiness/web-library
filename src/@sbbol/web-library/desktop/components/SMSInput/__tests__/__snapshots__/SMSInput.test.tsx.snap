// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SMSInput renders correctly 1`] = `
<div
  className="cssClass[globalSmsInput]"
  data-test-id="SMSInput"
>
  <div
    className="cssClass[btnRefresh]"
    onClick={[Function]}
  >
    <svg
      height="20"
      viewBox="0 0 20 20"
      width="20"
    >
      <clipPath
        id="clipFront0.999"
      >
        <path
          d="M10 10L21.883 8.33A12 12 0 1 1 21.099 5.438Z"
        />
      </clipPath>
      <path
        className="cssClass[light]"
        d="M18.1 9.101a.9.9 0 0 0-.9.9 7.245 7.245 0 1 1-1.998-4.95h-2.16a.9.9 0 1 0 0 1.8h4.077a.9.9 0 0 0 .9-.9v-4.05a.9.9 0 0 0-1.8 0v1.593A9 9 0 1 0 19 10.001a.9.9 0 0 0-.9-.9z"
      />
      <path
        className="cssClass[dark]"
        clipPath="url(#clipFront0.999)"
        d="M18.1 9.101a.9.9 0 0 0-.9.9 7.245 7.245 0 1 1-1.998-4.95h-2.16a.9.9 0 1 0 0 1.8h4.077a.9.9 0 0 0 .9-.9v-4.05a.9.9 0 0 0-1.8 0v1.593A9 9 0 1 0 19 10.001a.9.9 0 0 0-.9-.9z"
      />
    </svg>
  </div>
  <input
    autoComplete="off"
    className="cssClass[input]"
    maxLength={8}
    onChange={[Function]}
    onKeyDown={[Function]}
    placeholder="Введите код"
    type="text"
    value=""
  />
  <div
    className="cssClass[btnSubmit] cssClass[disabled]"
  >
    <svg
      className="svg-icon "
      height={20}
      viewBox="0 0 20 20"
      width={20}
    >
      <path
        d="M10.708 15.707l5-4.996a1 1 0 0 0 .21-.33c.1-.243.1-.516 0-.759a1 1 0 0 0-.21-.33l-5-4.996a1.005 1.005 0 0 0-1.42 1.42l3.3 3.286h-7.59a1 1 0 1 0 0 1.999h7.59l-3.3 3.287a.999.999 0 0 0 0 1.419 1 1 0 0 0 1.42 0z"
        fill="#FFF"
      />
    </svg>
  </div>
</div>
`;
